{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a76fa91e",
   "metadata": {},
   "source": [
    "# This code is used to extract data after mapping the pre-processed reads to the Virulence factor database (VFDB) with Bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3d69d724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Use this code for all the studies\n",
    "import pandas as pd\n",
    "import os\n",
    "import csv\n",
    "from tabulate import tabulate\n",
    "import re\n",
    "import shutil\n",
    "\n",
    "path='/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes/'\n",
    "os.chdir(path)\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe4e6f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a new folder to store all the Virulence factor genes data from all studies\n",
    "!mkdir /mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "studies = ['Belstrom_2021', 'Yost_2015', 'Jorth_2014', 'Duran_Pinedo_2014']\n",
    "destination_dir = '/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes'\n",
    "\n",
    "for study_name in studies:\n",
    "    source_dir = f\"/mnt/raid1/armen/Schinos/META-ANALYSIS/{study_name}/VirGenes\"\n",
    "\n",
    "    # Copy all files and directories from the VirGenes folder to the destination directory\n",
    "    for item in os.listdir(source_dir):\n",
    "        source_path = os.path.join(source_dir, item)\n",
    "        destination_path = os.path.join(destination_dir, item)\n",
    "\n",
    "        # Check if the item is a file before copying\n",
    "        if os.path.isfile(source_path):\n",
    "            shutil.copy2(source_path, destination_path)\n",
    "        else:\n",
    "            shutil.copytree(source_path, destination_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4c615f43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#open metadata file\n",
    "metadata = pd.read_csv(\"/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/RNA_metadata_ALL.tab\", sep= \"\\t\")\n",
    "#metadata.head()\n",
    "sample_names= metadata['External_ID'].tolist()\n",
    "#sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4068f9a1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Healthy IDs: ['Metatranscriptome_Healthy1', 'Metatranscriptome_Healthy2', 'Metatranscriptome_Healthy3', 'Metatranscriptome_Healthy4', 'Metatranscriptome_Healthy5', 'Metatranscriptome_Healthy6', 'RNA_12.1_S', 'RNA_13.1_S', 'RNA_15.1_S', 'RNA_21.1_S', 'RNA_22.1_S', 'RNA_28.1_S', 'RNA_8.1_S', 'RNA_9.1_S', 'SRR1045154', 'SRR1045155', 'SRR1045157', 'SRR13061493', 'SRR13061496', 'SRR13061499', 'SRR13061502', 'SRR13061592', 'SRR13061595', 'SRR13061598', 'SRR13061602', 'SRR13061606', 'SRR13061608']\n",
      "Periodontitis IDs: ['Metatranscriptome_Periodontitis1', 'Metatranscriptome_Periodontitis2', 'Metatranscriptome_Periodontitis3', 'Metatranscriptome_Periodontitis4', 'Metatranscriptome_Periodontitis5', 'Metatranscriptome_Periodontitis6', 'Metatranscriptome_Periodontitis7', 'RNA_12.2_P', 'RNA_15.2_P', 'RNA_21.2_P', 'RNA_22.3_P', 'RNA_26.2_P', 'RNA_28.3_P', 'RNA_8.2_P', 'RNA_9.2_P', 'SRR1045151', 'SRR1045152', 'SRR1045153', 'SRR13061556', 'SRR13061559', 'SRR13061562', 'SRR13061566', 'SRR13061569', 'SRR13061572', 'SRR13061578', 'SRR13061583', 'SRR13061586']\n"
     ]
    }
   ],
   "source": [
    "# Filter the DataFrame based on the 'Condition' column\n",
    "healthy_df = metadata.loc[metadata['Condition'] == 'Healthy']\n",
    "periodontitis_df = metadata.loc[metadata['Condition'] == 'Periodontitis']\n",
    "\n",
    "# Convert the 'External ID' column to a list for each filtered DataFrame\n",
    "healthy_ids = healthy_df['External_ID'].tolist()\n",
    "periodontitis_ids = periodontitis_df['External_ID'].tolist()\n",
    "\n",
    "# Display the lists of 'External ID' values\n",
    "print('Healthy IDs:', healthy_ids)\n",
    "print('Periodontitis IDs:', periodontitis_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ec2e7489",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy1/Metatranscriptome_Healthy1_vir_mapping.sam'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samfiles=[]\n",
    "for sample in sample_names:\n",
    "    samfile=path + sample +'/'+ sample +'_vir_mapping.sam'\n",
    "    samfiles.append(samfile)\n",
    "samfiles[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5ee317a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file /mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy1/Metatranscriptome_Healthy1_vir_mapping.sam\n",
      "samtools view: failed to open \"/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy1/Metatranscriptome_Healthy1_vir_mapping.sam\" for reading: No such file or directory\n",
      "[E::hts_open_format] Failed to open file /mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy2/Metatranscriptome_Healthy2_vir_mapping.sam\n",
      "samtools view: failed to open \"/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy2/Metatranscriptome_Healthy2_vir_mapping.sam\" for reading: No such file or directory\n",
      "[E::hts_open_format] Failed to open file /mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy3/Metatranscriptome_Healthy3_vir_mapping.sam\n",
      "samtools view: failed to open \"/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest/Metatranscriptome_Healthy3/Metatranscriptome_Healthy3_vir_mapping.sam\" for reading: No such file or directory\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### to be able to run bash command loop in python\n",
    "import subprocess\n",
    "# Loop \n",
    "for samfile in samfiles:\n",
    "    # Define the bash script to run for each element\n",
    "    samtools_script =f'samtools view -F 4 {samfile} | cut -f 3 | sort | uniq -c | sort -k1,1rn > /mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes/{os.path.basename(samfile)}_sorted_vir_genes.txt'\n",
    "    #This command filters the SAM file for reads that are mapped(-F 4) (if we want we can use -f 2 to filter the paired), \n",
    "    #and then extracts the reference name field (column 3) using the cut command. The output will be a \n",
    "    #list of reference names, one per line, for each mapped read. \n",
    "    #Finally, you can use the uniq -c command to count and then sort based on the counts of each reference name\n",
    "\n",
    "    # Use subprocess to run the bash script\n",
    "    subprocess.run(samtools_script, shell=True)    \n",
    "    \n",
    "#make a list of all files (sorted_vir_genes) in the directory VirulenceGenes that we created  \n",
    "# Get a list of all files and directories in the directory\n",
    "all_files = os.listdir(path)\n",
    "# Filter out the directories and hidden files to get only the file names\n",
    "file_names = [file for file in all_files if os.path.isfile(os.path.join(path, file)) and not file.startswith('.')]\n",
    "# Print the list of file names\n",
    "#print(file_names)   \n",
    "len(file_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bdc05854",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of virulence factor genes in the VFDB database is: 32439\n",
      "VFG037170(gb|WP_001081754) (plc1) phospholipase C [Phospholipase C (VF0470) - Exotoxin (VFC0235)] [Acinetobacter baumannii 1656-2]\n"
     ]
    }
   ],
   "source": [
    "#Save the names of each virulence factor genes from the VFDB in a list. Have a split version of it\n",
    "virulence_genes_fasta='/mnt/raid1/armen/Schinos/META-ANALYSIS/Databases/VFDB/VFDB_nt.fa'\n",
    "\n",
    "fasta_file = open(virulence_genes_fasta, \"r\")\n",
    "reference_names = []\n",
    "reference_split=[]\n",
    "for line in fasta_file:\n",
    "    if line.startswith(\">\"):\n",
    "        # Extract the reference name from the header line\n",
    "        reference_name = line[1:].strip()\n",
    "        reference_names.append(reference_name)\n",
    "        reference_split.append(reference_name.split())\n",
    "fasta_file.close()\n",
    "\n",
    "print('The total number of virulence factor genes in the VFDB database is: ' + str(len(reference_names)))\n",
    "print(reference_names[0])\n",
    "#print(len(reference_split))\n",
    "#print(reference_split[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6886fc4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the full name of the virulence gene found in the database\n",
    "with open('VirGenes_fullnames.txt', 'w') as file:\n",
    "    # Write each value on a new line separated by tabs\n",
    "    for value in reference_names:\n",
    "        file.write(value + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8b484963",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## For each sample, make a csv file with the Virulence Factor reference(NCBI gene accession),\n",
    "#Gene name, Encoded protein,Function, Origin and Counts\n",
    "\n",
    "\n",
    "#for each sample\n",
    "for f in file_names:\n",
    "    #open file\n",
    "    fl = open(path + f, \"r\")\n",
    "    read = fl.readlines()\n",
    "    fl.close()\n",
    "    vir_genes=[]\n",
    "    #strip each line and split in 2 parts(counts, reference) and store in the list 'lines'\n",
    "    vir_genes = [l.strip().split() for l in read]\n",
    "    #vir_genes[0]\n",
    "    \n",
    "    #a dictionary of identified virulence factor genes and their counts\n",
    "    vir_dict = {item[1]: item[0] for item in vir_genes}\n",
    "    #print(vir_dict)\n",
    "    \n",
    "    #a dictionary with the full reference name and their counts\n",
    "    items = list(vir_dict.items())\n",
    "    full_dict=dict()\n",
    "    for i in range(len(reference_split)):\n",
    "        for j in range(len(items)):\n",
    "            if reference_split[i][0]==items[j][0]:\n",
    "                full_dict[reference_names[i]]=items[j][1]\n",
    "\n",
    "    #Make a dictionary with the virulence factor genes sorted (decreasing) by the number of their counts\n",
    "    # Convert the values to integers\n",
    "    full_dict = {k: int(v) for k, v in full_dict.items()}\n",
    "    # Sort the dictionary based on the values in decreasing order\n",
    "    full_dict_sorted = dict(sorted(full_dict.items(), key=lambda item: item[1], reverse=True))\n",
    "    # Print the sorted dictionary\n",
    "    #print(len(full_dict_sorted))\n",
    "    #print(full_dict_sorted)\n",
    "    \n",
    "    #the patterns to check in the full reference name \n",
    "    pattern_gene = r'\\([^()]+\\).*?\\([^()]+\\)\\s+(.+?)\\s*\\['\n",
    "    # In the above pattern, we first match any sequence of characters that is enclosed in parentheses '\\([^()]+\\)'. \n",
    "    # This matches the first set of parentheses e.g. (gb|WP_006154814).\n",
    "    # We then match any sequence of characters non-greedily, i.e., matching the shortest possible sequence of characters,\n",
    "    # using '.*?'. This matches the characters between the first set of parentheses and the second set of parentheses.\n",
    "    # Next, we match the second set of parentheses \\([^()]+\\) and one or more spaces \\s+, \n",
    "    # followed by the substring you want (.+?), which is the shortest possible sequence of one or more characters ending \n",
    "    # with a space. Finally, we match zero or more spaces \\s* followed by the character '['.\n",
    "\n",
    "    pattern_function = r'(?<=\\s-\\s)[\\w\\s/]+(?=\\()'\n",
    "    #This pattern uses a positive lookahead (?=\\() to match all the words after '-' and before the next '('. \n",
    "    #The [\\w\\s/]+ matches one or more word characters and spaces (including '/'). \n",
    "    #The pattern is surrounded by the lookbehind (?<=\\s-\\s) and lookahead (?=\\() assertions.\n",
    "\n",
    "    pattern_species = r'\\[([^\\[\\]]+)\\][^\\[\\]]*\\['\n",
    "    # In this pattern, we use square brackets \\[ and \\] to match literal opening and closing brackets, respectively. The \n",
    "    # ([^\\[\\]]+) pattern matches one or more characters that are not brackets and captures them in a group.\n",
    "    # The [^\\[\\]]*\\[ pattern matches any characters that are not brackets\n",
    "    # and followed by an opening bracket. This ensures that we are selecting the second pair of brackets in the string.\n",
    "\n",
    "    # Convert the dictionary to a list of tuples\n",
    "    table_data = [(key.split()[0][:10] + key.split()[0][13:], key.split()[1].strip('()'), \\\n",
    "                   re.search(pattern_gene, key).group(1) if re.search(pattern_gene, key) else '', \\\n",
    "                   re.search(pattern_function, key).group(0) if re.search(pattern_function, key) else '', \\\n",
    "                   re.search(pattern_species, key).group(1) if re.search(pattern_species, key) else '', value)\\\n",
    "                   for key, value in full_dict_sorted.items()]\n",
    "\n",
    "    # Print the table\n",
    "    #print(tabulate(table_data, headers=['Virulence Factor reference(NCBI gene accession)','Gene name', 'Gene',\\\n",
    "    #                                    'Function', 'Species', 'Counts']))\n",
    "    \n",
    "    # For the proper naming of the files\n",
    "    # Find the starting index of the desired substring\n",
    "    start_index = f.find(\"_vir_mapping.sam_sorted_vir_genes.txt\")\n",
    "    # Extract the desired substring\n",
    "    desired_string = f[:start_index]\n",
    "    # Create the file name using the desired substring\n",
    "    file_name = desired_string + \"_results\"\n",
    "\n",
    "    \n",
    "    # Save the table as a tab-delimited file\n",
    "    with open(file_name, \"w\", newline=\"\") as file:\n",
    "        writer = csv.writer(file, delimiter=\"\\t\")\n",
    "        writer.writerow(['Virulence Factor reference(NCBI gene accession)','Gene name', 'Encoded protein', 'Function',\\\n",
    "                     'Origin', 'Counts'])\n",
    "        for row in table_data:\n",
    "            writer.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f4913288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
      "/tmp/ipykernel_1723648/2206396046.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "#Here we extract the 'Virulence Factor reference(NCBI gene accession)' and 'Counts' columns from \n",
    "# each dataframe. We then rename the 'Counts' column with the respective dataframe name. Finally, we merge the \n",
    "# dataframes based on the 'Virulence Factor reference(NCBI gene accession)' column and set it as the row index \n",
    "# in the merged dataframe.\n",
    "\n",
    "# Create an empty list to store the DataFrames\n",
    "df_healthy = []\n",
    "df_peri=[]\n",
    "\n",
    "# Loop through the healthy and periodontitis samples and append to a list of dataframes\n",
    "for sample in healthy_ids:\n",
    "    df = pd.read_csv(sample + \"_results\", sep='\\t')\n",
    "    df_healthy.append(df)\n",
    "\n",
    "for sample in periodontitis_ids:\n",
    "    df = pd.read_csv(sample + \"_results\", sep='\\t')\n",
    "    df_peri.append(df)    \n",
    "\n",
    "\n",
    "# Create an empty dataframe to store the merged data\n",
    "merged_healthy = pd.DataFrame()\n",
    "merged_peri = pd.DataFrame()\n",
    "\n",
    "\n",
    "# Iterate over the dataframes in the list\n",
    "#helathy\n",
    "for i, df in enumerate(df_healthy):\n",
    "    # Extract the 'Virulence Factor reference(NCBI gene accession)' and 'Counts' columns\n",
    "    df_counts = df[['Virulence Factor reference(NCBI gene accession)', 'Counts']]\n",
    "    \n",
    "    # Rename the 'Counts' column with the dataframe name\n",
    "    df_counts.rename(columns={'Counts': healthy_ids[i]}, inplace=True)\n",
    "    \n",
    "    # Merge the current dataframe with the merged dataframe\n",
    "    if merged_healthy.empty:\n",
    "        merged_healthy = df_counts\n",
    "    else:\n",
    "        merged_healthy = merged_healthy.merge(df_counts, how='outer', on='Virulence Factor reference(NCBI gene accession)')\n",
    "\n",
    "#Periodontitis\n",
    "for i, df in enumerate(df_peri):\n",
    "    # Extract the 'Virulence Factor reference(NCBI gene accession)' and 'Counts' columns\n",
    "    df_counts = df[['Virulence Factor reference(NCBI gene accession)', 'Counts']]\n",
    "    \n",
    "    # Rename the 'Counts' column with the dataframe name\n",
    "    df_counts.rename(columns={'Counts': periodontitis_ids[i]}, inplace=True)\n",
    "    \n",
    "    # Merge the current dataframe with the merged dataframe\n",
    "    if merged_peri.empty:\n",
    "        merged_peri = df_counts\n",
    "    else:\n",
    "        merged_peri = merged_peri.merge(df_counts, how='outer', on='Virulence Factor reference(NCBI gene accession)')\n",
    "\n",
    "        \n",
    "        \n",
    "# Set the 'Virulence Factor reference(NCBI gene accession)' as the row index\n",
    "#healthy\n",
    "merged_healthy.set_index('Virulence Factor reference(NCBI gene accession)', inplace=True)\n",
    "#periodontitis\n",
    "merged_peri.set_index('Virulence Factor reference(NCBI gene accession)', inplace=True)\n",
    "\n",
    "#merge healthy and periodontitis dataframes together\n",
    "merged_ALL = pd.merge(merged_healthy, merged_peri, on='Virulence Factor reference(NCBI gene accession)')\n",
    "# Fill all NaN values with zeros\n",
    "df_filled = merged_ALL.fillna(0)\n",
    "\n",
    "# Save the merged dataframes as a tab-delimited CSV file\n",
    "merged_healthy.to_csv('VirGenes_healthy.csv', sep='\\t', index=True)\n",
    "merged_peri.to_csv('VirGenes_peri.csv', sep='\\t', index=True)\n",
    "df_filled.to_csv('VirGenes_ALL.csv', sep='\\t')\n",
    "\n",
    "\n",
    "# Print the merged dataframe\n",
    "#print(merged_healthy)\n",
    "#print(merged_peri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1bbfba47",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenesTest\n"
     ]
    }
   ],
   "source": [
    "# current_dir \n",
    "print(os.getcwd())\n",
    "\n",
    "# Change the current working directory\n",
    "path = \"/mnt/raid1/armen/Schinos/META-ANALYSIS/ALL/Fungi/VirulenceGenes/\"\n",
    "os.chdir(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "523d5d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virulence Factor reference(NCBI gene accession)</th>\n",
       "      <th>Metatranscriptome_Healthy4</th>\n",
       "      <th>Metatranscriptome_Healthy5</th>\n",
       "      <th>Metatranscriptome_Healthy6</th>\n",
       "      <th>RNA_12.1_S</th>\n",
       "      <th>RNA_13.1_S</th>\n",
       "      <th>RNA_15.1_S</th>\n",
       "      <th>RNA_21.1_S</th>\n",
       "      <th>RNA_22.1_S</th>\n",
       "      <th>RNA_28.1_S</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR1045153</th>\n",
       "      <th>SRR13061556</th>\n",
       "      <th>SRR13061559</th>\n",
       "      <th>SRR13061562</th>\n",
       "      <th>SRR13061566</th>\n",
       "      <th>SRR13061569</th>\n",
       "      <th>SRR13061572</th>\n",
       "      <th>SRR13061578</th>\n",
       "      <th>SRR13061583</th>\n",
       "      <th>SRR13061586</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VFG005359(WP_002899102)</td>\n",
       "      <td>314.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>244.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>647.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3967.0</td>\n",
       "      <td>570.0</td>\n",
       "      <td>334.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>322.0</td>\n",
       "      <td>2470.0</td>\n",
       "      <td>690.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>802.0</td>\n",
       "      <td>764.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VFG005810(WP_011836396)</td>\n",
       "      <td>163.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>...</td>\n",
       "      <td>498.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>246.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VFG043592(AAF34780)</td>\n",
       "      <td>149.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>307.0</td>\n",
       "      <td>...</td>\n",
       "      <td>116.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>1348.0</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VFG043604(WP_008810020)</td>\n",
       "      <td>148.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>1774.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>401.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1252.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1017.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>649.0</td>\n",
       "      <td>513.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VFG005582(WP_002897814)</td>\n",
       "      <td>137.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1986.0</td>\n",
       "      <td>494.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>523.0</td>\n",
       "      <td>368.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>310.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Virulence Factor reference(NCBI gene accession)  Metatranscriptome_Healthy4  \\\n",
       "0                         VFG005359(WP_002899102)                       314.0   \n",
       "1                         VFG005810(WP_011836396)                       163.0   \n",
       "2                             VFG043592(AAF34780)                       149.0   \n",
       "3                         VFG043604(WP_008810020)                       148.0   \n",
       "4                         VFG005582(WP_002897814)                       137.0   \n",
       "\n",
       "   Metatranscriptome_Healthy5  Metatranscriptome_Healthy6  RNA_12.1_S  \\\n",
       "0                        15.0                       172.0       145.0   \n",
       "1                         2.0                        17.0        21.0   \n",
       "2                        10.0                       139.0         0.0   \n",
       "3                         0.0                        44.0        61.0   \n",
       "4                         4.0                        87.0        51.0   \n",
       "\n",
       "   RNA_13.1_S  RNA_15.1_S  RNA_21.1_S  RNA_22.1_S  RNA_28.1_S  ...  \\\n",
       "0       244.0       547.0        10.0       266.0       647.0  ...   \n",
       "1        30.0        21.0         2.0       127.0        35.0  ...   \n",
       "2        52.0       172.0         2.0         0.0       307.0  ...   \n",
       "3        81.0      1774.0        15.0        26.0       401.0  ...   \n",
       "4        60.0       362.0         0.0       170.0       210.0  ...   \n",
       "\n",
       "   SRR1045153  SRR13061556  SRR13061559  SRR13061562  SRR13061566  \\\n",
       "0      3967.0        570.0        334.0         59.0        322.0   \n",
       "1       498.0         44.0        127.0          0.0         65.0   \n",
       "2       116.0        120.0          4.0         22.0          0.0   \n",
       "3         3.0       1252.0         22.0         73.0          0.0   \n",
       "4      1986.0        494.0        142.0         31.0        123.0   \n",
       "\n",
       "   SRR13061569  SRR13061572  SRR13061578  SRR13061583  SRR13061586  \n",
       "0       2470.0        690.0         47.0        802.0        764.0  \n",
       "1        246.0         57.0          4.0         15.0         20.0  \n",
       "2          0.0       1178.0        127.0       1348.0        114.0  \n",
       "3       1017.0         27.0         90.0        649.0        513.0  \n",
       "4        523.0        368.0         39.0        284.0        310.0  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the tab-delimited file into a DataFrame\n",
    "VirGenes = pd.read_csv(path + \"VirGenes_ALL.csv\", delimiter='\\t')\n",
    "\n",
    "# Display the DataFrame\n",
    "VirGenes.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e44e003c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(802, 52)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VirGenes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bc3d32ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metatranscriptome_Healthy4</th>\n",
       "      <th>Metatranscriptome_Healthy5</th>\n",
       "      <th>Metatranscriptome_Healthy6</th>\n",
       "      <th>RNA_12.1_S</th>\n",
       "      <th>RNA_13.1_S</th>\n",
       "      <th>RNA_15.1_S</th>\n",
       "      <th>RNA_21.1_S</th>\n",
       "      <th>RNA_22.1_S</th>\n",
       "      <th>RNA_28.1_S</th>\n",
       "      <th>RNA_8.1_S</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR1045153</th>\n",
       "      <th>SRR13061556</th>\n",
       "      <th>SRR13061559</th>\n",
       "      <th>SRR13061562</th>\n",
       "      <th>SRR13061566</th>\n",
       "      <th>SRR13061569</th>\n",
       "      <th>SRR13061572</th>\n",
       "      <th>SRR13061578</th>\n",
       "      <th>SRR13061583</th>\n",
       "      <th>SRR13061586</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Common reference</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAB30711</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC82625</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3276.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAD53408</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAF34780</th>\n",
       "      <td>149.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>307.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>116.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>1348.0</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAF37887</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metatranscriptome_Healthy4  Metatranscriptome_Healthy5  \\\n",
       "Common reference                                                           \n",
       "AAB30711                                 0.0                         0.0   \n",
       "AAC82625                                 0.0                         0.0   \n",
       "AAD53408                                 0.0                         0.0   \n",
       "AAF34780                               149.0                        10.0   \n",
       "AAF37887                                 0.0                         3.0   \n",
       "\n",
       "                  Metatranscriptome_Healthy6  RNA_12.1_S  RNA_13.1_S  \\\n",
       "Common reference                                                       \n",
       "AAB30711                                 0.0         0.0         0.0   \n",
       "AAC82625                                 0.0         0.0         0.0   \n",
       "AAD53408                                 0.0         0.0         0.0   \n",
       "AAF34780                               139.0         0.0        52.0   \n",
       "AAF37887                                 0.0         0.0         0.0   \n",
       "\n",
       "                  RNA_15.1_S  RNA_21.1_S  RNA_22.1_S  RNA_28.1_S  RNA_8.1_S  \\\n",
       "Common reference                                                              \n",
       "AAB30711                 0.0         0.0         0.0         0.0        0.0   \n",
       "AAC82625                 0.0        20.0         0.0         0.0        0.0   \n",
       "AAD53408                 0.0         0.0         0.0         0.0        0.0   \n",
       "AAF34780               172.0         2.0         0.0       307.0        0.0   \n",
       "AAF37887                 0.0         0.0         0.0         0.0        0.0   \n",
       "\n",
       "                  ...  SRR1045153  SRR13061556  SRR13061559  SRR13061562  \\\n",
       "Common reference  ...                                                      \n",
       "AAB30711          ...         0.0          8.0          0.0         40.0   \n",
       "AAC82625          ...      3276.0          0.0          6.0         16.0   \n",
       "AAD53408          ...         6.0          0.0          0.0          0.0   \n",
       "AAF34780          ...       116.0        120.0          4.0         22.0   \n",
       "AAF37887          ...         0.0          0.0          0.0          0.0   \n",
       "\n",
       "                  SRR13061566  SRR13061569  SRR13061572  SRR13061578  \\\n",
       "Common reference                                                       \n",
       "AAB30711                  0.0          0.0          0.0          4.0   \n",
       "AAC82625                  9.0          2.0          3.0         26.0   \n",
       "AAD53408                  0.0          0.0          0.0          0.0   \n",
       "AAF34780                  0.0          0.0       1178.0        127.0   \n",
       "AAF37887                  0.0          0.0          0.0          0.0   \n",
       "\n",
       "                  SRR13061583  SRR13061586  \n",
       "Common reference                            \n",
       "AAB30711                  0.0          0.0  \n",
       "AAC82625                 13.0          4.0  \n",
       "AAD53408                  0.0          0.0  \n",
       "AAF34780               1348.0        114.0  \n",
       "AAF37887                  0.0          0.0  \n",
       "\n",
       "[5 rows x 51 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the column containing the values with substrings between parentheses\n",
    "#column_name = 'Virulence Factor reference(NCBI gene accession)'\n",
    "\n",
    "# Extract the common substring between parentheses for each value\n",
    "VirGenes['Common reference'] = VirGenes['Virulence Factor reference(NCBI gene accession)'].str.extract(r'\\((.*?)\\)')\n",
    "\n",
    "# Group the DataFrame by the common substring and merge the columns by summing their values\n",
    "merged_df = VirGenes.groupby('Common reference').sum()\n",
    "\n",
    "# Save as a tab-delimited file\n",
    "merged_df.to_csv('VirGenes_All_merged.csv', sep='\\t', index=True)\n",
    "\n",
    "# Remove the 'Common Substring' column if not needed in the final result\n",
    "#merged_df.drop('Common Substring', axis=1, inplace=True)\n",
    "\n",
    "# Display the merged DataFrame\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a25d3c67",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
