---
title: "mrna_counts_analysis"
output:
  word_document: default
  html_document:
    df_print: paged
    fig_caption: yes
    self_contained: no
params:
  output_folder: 
    value: "/mnt/raid1/philip/schinos_seq/results/mrna"
date: '2024-02-23'
---

```{r setup, include=FALSE}
#rm(list = ls())
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE, out.width="100%", out.height="100%")
knitr::opts_chunk$set(dpi=400,fig.width=9, fig.height=9)
output_folder = params$output_folder

```

# mrna analysis, SCHINOS {.tabset .tabset-fade .tabset-pills}

```{r libraries}
# BiocManager::install(version = "3.16")
#library("BiocManager")
# BiocManager::install("edgeR")
library(edgeR)
# BiocManager::install("tximportData")
library(tximportData)
#library(tximport)
library(dplyr)
library(stringr)
library(RColorBrewer)
library(readxl)
library(org.Hs.eg.db)
library(ggplot2)
library(knitr)
```

```{r read_files}
dir <- "/mnt/raid1/philip/schinos_seq/data/mrna/mapped_to_genes"
files = list.files(dir)[grepl("ReadsPerGene", list.files(dir))]
names(files) <- gsub("_S.*ReadsPerGene.*","",files)
files_paths=file.path(dir,files)
names(files_paths) <- names(files)

#####read transcript counts from STAR
# dir <- "/mnt/raid1/philip/schinos_seq/data/mrna/counts"
# files_names = list.files(dir)[grepl(".*genes\\.results", list.files(dir))]
# names(files_names) <- gsub("_S.*results","",files_names)
# files_paths=file.path(dir,files_names)
# names(files_paths) <- names(files_names)
#txi.rsem <- tximport(files_paths, type = "rsem", txIn = FALSE, txOut = FALSE)
#depth = colSums(txi.rsem$counts)
# #re-arrange order of samples
#depth=sort(depth, decreasing=TRUE)
#depth = depth[c(grep("TI", names(depth)),grep("GF", names(depth)))]

######read gene counts from STAR

counts_table = sapply(files_paths, function(i)
{
  #the 10th column has the counts
  x = read.csv(i, sep="\t", stringsAsFactors = FALSE, header = FALSE, skip=4)[, 4]
})
rownames(counts_table) = rownames(read.csv(files_paths[1], sep="\t", stringsAsFactors = FALSE, header = FALSE, row.names = 1, skip=4))

depth = colSums(counts_table)
# #re-arrange order of samples
depth=sort(depth, decreasing=TRUE)
depth = depth[c(grep("TI", names(depth)),grep("GF", names(depth)))]
#PLOT
barplot(depth/1e06, names=colnames(depth),las=2, cex.names = 0.75); grid()
#axis(side=2, at=)
mtext(side = 2, text = "Library size (millions)", line = 3)
title("Barplot of sum of gene counts")
```
```{r read_metadata}
metadata_file="/mnt/raid1b/philip/schinos_seq/metadata/SCHINOS_Sequenced_samples_metadata.tsv"
metadata = read.csv(metadata_file, sep="\t", stringsAsFactors = FALSE)
rownames(metadata) <- metadata$Sample_name
metadata$Timepoint=factor(metadata$Timepoint)
metadata$Mastic_product=relevel(factor(metadata$Mastic_product),"Placebo gum")
metadata = metadata %>% filter(Discarded==0,na.rm=TRUE)
metadata = metadata %>% filter(str_detect(Sample_name,"MRN"),na.rm=TRUE)
metadata$Batch=factor(metadata$Batch)
#change order of metadata rows to match files list
metadata=metadata[match(names(files_paths),metadata$Sample_name),]
```



```{r edgeR}
y <- DGEList(counts_table, group = metadata$Mastic_product)
y$samples=merge(y$samples,metadata, by="row.names")

#annotate gene ids
gene_info=as.data.frame ( read.table(file = "/mnt/raid1b/philip/Homo_sapiens/STAR_index/geneInfo.tab", header=F, skip=1, sep="\t"))
colnames(gene_info) = c("ENSEMBL","SYMBOL","GENETYPE")
gene_info$GENENAME = mapIds(org.Hs.eg.db, keys=gene_info$ENSEMBL,
                   keytype="ENSEMBL", column="GENENAME")
geneids=data.frame(ENSEMBL = rownames(y) )
y$genes = inner_join(geneids, gene_info, by="ENSEMBL")

#save counts table
table_to_print = as.data.frame(y$counts)
table_to_print = tibble::rownames_to_column(table_to_print, var="ENSEMBL")
write.table(table_to_print,file=file.path(output_folder,"mrna_counts.tsv"), sep="\t", quote=FALSE, row.names = FALSE)
#save RPMs table
table_to_print = as.data.frame(round(cpm(y),digits=2))
table_to_print$SYMBOL = gene_info$SYMBOL[match(rownames(table_to_print),gene_info$ENSEMBL )]
table_to_print = tibble::rownames_to_column(table_to_print, var="ENSEMBL")
#put SYMBOL COLUMN SECOND
table_to_print = table_to_print[ , c(1, ncol(table_to_print),3:ncol(table_to_print)-1)]
write.table(table_to_print,file=file.path(output_folder,"mrna_rpms.tsv"), sep="\t", quote=FALSE, row.names = FALSE)
#save gene_info_table
write.table(gene_info,file=file.path(output_folder,"genes_info.tsv"), sep="\t", quote=FALSE, row.names = FALSE)

#Filtering low counts
# keep.exprs <- filterByExpr(y, group=group, min.count=5) # min.count = 10 read in the median library size ~10/2,200,000 = 4.5 CPM, min.total.count = 15, in 0.7*18=13 samples (13=70% of smallest group size)
# y <- y[keep.exprs,, keep.lib.sizes=FALSE]  #filter and recalculate library size
#dim(y)

#gingival crevicular fluid samples only
y_GF = y[,which(y$samples$Sample_type_abbreviation=="GF")]   
#tissue swab samples only
y_TI = y[,which(y$samples$Sample_type_abbreviation=="TI")]  


#####SAVE GF ONLY
table_to_print = as.data.frame(y_GF$counts)
table_to_print = tibble::rownames_to_column(table_to_print, var="ENSEMBL")
write.table(table_to_print,file=file.path(output_folder,"mrna_counts_GF.tsv"), sep="\t", quote=FALSE, row.names = FALSE)
#save RPMs table of GF samples only
table_to_print = as.data.frame(round(cpm(y_GF),digits=2))
table_to_print$SYMBOL = gene_info$SYMBOL[match(rownames(table_to_print),gene_info$ENSEMBL )]
table_to_print$GENETYPE = gene_info$GENETYPE[match(rownames(table_to_print),gene_info$ENSEMBL )]
table_to_print$GENENAME = gene_info$GENENAME[match(rownames(table_to_print),gene_info$ENSEMBL )]
table_to_print = tibble::rownames_to_column(table_to_print, var="ENSEMBL")
#put SYMBOL COLUMN SECOND etc
columns = ncol(table_to_print)
table_to_print = table_to_print[ , c(1, columns-2, columns-1, columns, 5:columns-3)]
write.table(table_to_print,file=file.path(output_folder,"mrna_rpms_GF_with_geneinfo.tsv"), sep="\t", quote=FALSE, row.names = FALSE)
### save only genes with sum cpm>=540 (mean >=10)
means = cpm(y_GF) %>% rowMeans()
write.table(table_to_print[means>=10,-c(3,4)],file=file.path(output_folder,"mrna_rpms_GF_filtered_with_SYMBOL.tsv"), sep="\t", quote=FALSE, row.names = FALSE)

```

## Plots

```{r plot_before_filtering_genes, fig.height=4}
y=y_GF
L <- mean(y$samples$lib.size) * 1e-6
M <- median(y$samples$lib.size) * 1e-6

# 2.925283 2.511230
cpm <- cpm(y)
lcpm <- cpm(y, log=TRUE)  #lcpm is calculated as log2(cpm +2/L) to ensure non-zeroes

col <- rep(brewer.pal(12, "Paired"),6)
boxplot(lcpm, las=2, col=col, main="", cex=0.7, cex.axis=0.6)
title(main="A. Unnormalised unfiltered data (expressed in log-CPM)", ylab="Log-cpm")
```
```{r}
#choose minimum cutoff
counts.cutoff = 10
lcpm.cutoff <- log2(counts.cutoff/M + 2/L)
##### because lcpm = log2(rawcount/M +2/L)

nsamples <- ncol(y)
samplenames=colnames(y)

#par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n",ncol=2, cex=0.8)

```

## Filtering and normalisation

```{r filtering_and_plot}
y=y_GF
y$samples$Batch = factor(y$samples$Batch)
libsizes = y$samples$lib.size
#keep samples with depth >=1 million
y = y[,which(y$samples$lib.size>1000000)]   
#keep genes that are present in at least 18 samples with at least 10 reads
keep.exprs <- filterByExpr(y, group=y$samples$group, min.count=counts.cutoff, large.n=18, min.prop=1)
y <- y[keep.exprs,, keep.lib.sizes=FALSE]   #change to FALSE
dim(y)
lcpm <- cpm(y, log=TRUE)
nsamples <- ncol(y)
samplenames=colnames(y)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n", ncol=2,cex=0.8)

```
```{r Normalising_gene_expression_distributions, fig.height=4}
#y$samples$norm.factors
#par(mfrow=c(1,2))
lcpm <- cpm(y, log=TRUE)
#png(filename=file.path(output_folder,"plots","B. Unnormalised filtered data"),)
bp = boxplot(lcpm, las=2, col=col, main="", cex=0.7, cex.axis=0.6)
title(main="B. Unnormalised filtered data", ylab="Log-cpm")
#dev.off()

y <- calcNormFactors(y, method = "TMM")
lcpm <- cpm(y, log=TRUE)
#png(filename=file.path(output_folder,"plots","C. Normalised filtered data.png"))
bp = boxplot(lcpm, las=2, col=col, main="", cex=0.7, cex.axis=0.6, fig.height=1)
title(main="C. Normalised filtered data", ylab="Log-cpm")
#dev.off()

```

```{r plot_MDS_and_then_fit, out.width='50%'}
#exploring differences between libraries MDS plot
pch <- c(15,16,17)
colors=c("darkgrey","#4588e6","#67ab43")
plotMDS(y, pch = as.numeric(y$samples$Mastic_product), col=colors[as.numeric(y$samples$Mastic_product)], labels =gsub("MRN","",y$samples$Sample_name), cex=0.8)
legend("topleft", legend=levels(y$samples$group), pch = pch, col= colors[c(1:nlevels(y$samples$Mastic_product))],cex=1)
title("All gingival crevicular fluid samples")

###day 0
y_filtered = y[,which(y$samples$Timepoint == 1)]
plotMDS(y_filtered , pch = as.numeric(y_filtered$samples$Mastic_product), col=colors[as.numeric(y_filtered$samples$Mastic_product)], labels =gsub("MRN","",y_filtered$samples$Sample_name), cex=1)
legend("topleft", legend=levels(y_filtered$samples$group), pch = pch, col= colors[c(1:nlevels(y_filtered$samples$Mastic_product))],cex=1)
title("All Day 0 samples")

y_pg=y[,which(y$samples$Mastic_product=="Placebo gum")]   
y_mg=y[,which(y$samples$Mastic_product=="Mastic gum")]   
y_mo=y[,which(y$samples$Mastic_product=="Mastic oil")] 

for(group in c("all", "Placebo gum", "Mastic gum", "Mastic oil")){
  pch = 15
  if ( group == "all") {
    y_filtered = y
  }else{
      y_filtered=y[,which(y$samples$Mastic_product==group)] 
  }
  par(mar=c(5,5,5,7))
  colors=c("darkred","red","pink")
  plotMDS(y_filtered , pch=pch, col=colors[as.numeric(y_filtered$samples$Timepoint)], labels =gsub("GF..MRN","",y_filtered$samples$Sample_name), cex=1.1)
  legend("topleft", inset=c(-0,0), xpd = TRUE, legend=paste0("Timepoint ", levels(y_filtered$samples$Timepoint)), pch = pch, col= colors[c(1:nlevels(y_filtered$samples$Timepoint))],cex=1.1)
  title(paste0(group," Patients"))
}

#plot batch effect
y_filtered= y
pch=15
colors = c("red","green","blue","orange")
plotMDS(y_filtered , pch = as.numeric(y_filtered$samples$Mastic_product), col=colors[as.numeric(y_filtered$samples$Batch)], labels =gsub("MRN","",y_filtered$samples$Sample_name), cex=1)
legend("topleft", legend=paste0("Batch ",levels(y_filtered$samples$Batch)), pch = pch, col= colors[c(1:nlevels(y_filtered$samples$Batch))],cex=1)
title("All samples")


# plotBCV(y)
# design <- model.matrix(~0+y$samples$Timepoint)
# colnames(design) <- gsub("y.*samples\\$","",colnames(design))
# y_fit <- glmQLFit(y, design, robust=TRUE)
# plotQLDisp(y_fit)

#design matrix
  
```

## Mastic gum

```{r mastic_gum}
#MASTIC GUM GROUP
y_current = y_mg
design <- model.matrix(~0+y_current$samples$Timepoint+y_current$samples$Batch)
colnames(design) <- gsub("y.*samples\\$","",colnames(design))

#Dispersion estimation
y_current <- estimateDisp(y_current, design, robust=TRUE)
#plotBCV(y_current)
y_fit <- glmQLFit(y_current, design, robust=TRUE)
#plotQLDisp(y_fit)

t2.vs.t1 <- makeContrasts(Timepoint2 - Timepoint1, levels = design) 
result <- glmQLFTest(y_fit, contrast=t2.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t2.vs.t1_summary_batch.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t2.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)

t3.vs.t1 <- makeContrasts(Timepoint3 - Timepoint1, levels = design) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t3.vs.t1_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t3.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
######################################################
t3.vs.t2 <- makeContrasts(Timepoint3 - Timepoint2, levels = design) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t2)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t3.vs.t2_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t3.vs.t2.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
######################################################

```

## Mastic oil

```{r mastic_oil_group}
#MASTIC OIL GROUP
design_mo <- model.matrix(~0+y_mo$samples$Timepoint)
colnames(design_mo) <- gsub("y.*samples\\$","",colnames(design_mo))

#Dispersion estimation
y_mo <- estimateDisp(y_mo, design_mo, robust=TRUE)
#plotBCV(y_mo)
y_fit <- glmQLFit(y_mo, design_mo, robust=TRUE)
#plotQLDisp(y_fit)

t2.vs.t1 <- makeContrasts(Timepoint2 - Timepoint1, levels = design_mo) 
result <- glmQLFTest(y_fit, contrast=t2.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t2.vs.t1_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t2.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)


t3.vs.t1 <- makeContrasts(Timepoint3 - Timepoint1, levels = design_mo) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t3.vs.t1_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t3.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
###
t3.vs.t2 <- makeContrasts(Timepoint3 - Timepoint2, levels = design_mo) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t2)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t3.vs.t2_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mo_t3.vs.t2.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
######################################################

```

```{r placebo_gum_group}
#placebo GROUP
design_pg <- model.matrix(~0+y_pg$samples$Timepoint)
colnames(design_pg) <- gsub("y.*samples\\$","",colnames(design_pg))

#Dispersion estimation
y_pg <- estimateDisp(y_pg, design_pg, robust=TRUE)
#plotBCV(y_pg)
colnames(design_pg) <- gsub("y.*samples\\$","",colnames(design_pg))
y_fit <- glmQLFit(y_pg, design_mo, robust=TRUE)
#plotQLDisp(y_fit)

t2.vs.t1 <- makeContrasts(Timepoint2 - Timepoint1, levels = design_pg) 
result <- glmQLFTest(y_fit, contrast=t2.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t2.vs.t1_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t2.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)


t3.vs.t1 <- makeContrasts(Timepoint3 - Timepoint1, levels = design_pg) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t1)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t3.vs.t1_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t3.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
p <- ggplot(data=result_table, aes(x=logFC, y=FDR)) + geom_point() +
  geom_hline(yintercept=1, linetype="dashed", color = "red")+ geom_vline(xintercept=c(-1,1), linetype="dashed", color = "red")
p


t3.vs.t2 <- makeContrasts(Timepoint3 - Timepoint2, levels = design_pg) 
result <- glmQLFTest(y_fit, contrast=t3.vs.t2)
is.de <- decideTestsDGE(result)
summary(is.de)
write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t3.vs.t2_summary.tsv", sep ="\t", quote = FALSE)
result_table = topTags(result, n = Inf)$table
write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_pg_t3.vs.t2.tsv", sep ="\t", quote = FALSE, row.names = FALSE)
p <- ggplot(data=result_table, aes(x=logFC, y=-FDR, name=SYMBOL)) + geom_point() + #geom_text(aes(label=SYMBOL))+
  geom_hline(yintercept=-0.05, linetype="dashed", color = "red")+ geom_vline(xintercept=c(-1,1), linetype="dashed", color = "red")
p
```
```{r high_vs_low_bop}
samples_high = c(
  "G2GF00MRN",
  "G7GF00MRN",
  "S9GF00MRN",
  "G1GF00MRN",
  "S8GF00MRN",
  "G3GF00MRN",
  "S5GF00MRN",
  "S11GF00MRN",
  "S2GF00MRN",
  "G13GF00MRN")


samples_low=c(  
  "G12GF07MRN",
  "G11GF07MRN",
  "G3GF21MRN",
  "G8GF07MRN",
  "G7GF21MRN",
  "G6GF07MRN",
  "G13GF07MRN",
  "G5GF07MRN",
  "G7GF07MRN",
  "S6GF07MRN")
  

y_current = y
y_current  = y_current[,which(y_current$samples$Sample_name %in% c(samples_low,samples_high))]

y_current$samples$BoPgroup = NA
y_current$samples$BoPgroup[y_current$samples$Sample_name %in% samples_low]="low"
y_current$samples$BoPgroup[y_current$samples$Sample_name %in% samples_high]="high"

#y$samples$Batch = factor(y$samples$Batch)
design <- model.matrix(~0+y_current$samples$BoPgroup)
colnames(design) <- gsub("y.*samples\\$","",colnames(design))

#Dispersion estimation
y_current <- estimateDisp(y_current, design, robust=TRUE)
#plotBCV(y_current)
y_fit <- glmQLFit(y_current, design, robust=TRUE)
#plotQLDisp(y_fit)



h.vs.l <- makeContrasts(BoPgrouphigh - BoPgrouplow, levels = design) 
result <- glmQLFTest(y_fit, contrast=h.vs.l)
is.de <- decideTestsDGE(result)
summary(is.de)
#write.table(summary(is.de), file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t2.vs.t1_summary_batch.tsv", sep ="\t", quote = FALSE)
#result_table = topTags(result, n = Inf)$table
#write.table(result_table, file = "/mnt/raid1/philip/schinos_seq/results/mrna/DE/edgeR_mg_t2.vs.t1.tsv", sep ="\t", quote = FALSE, row.names = FALSE)



```